### DISCUSSION TITLE
Error reading data in Firestore with Nuxt Js using Fetch Hook

### DISCUSSION THREAD

**Discussion by tomifis** [nuxt/nuxt#17969] [Questions]
I was trying to do pagination with firebase and nuxt js . When trying to read from firestore using the fetch hook I get an error (Maximum call stack size exceeded ). It is worth noting that when doing this same process with the hook mounted works perfectly. I need to use the fetch hook for ssr. I think the error comes from assigning the value to the variable lastDoc which I need to save the cursor that I will later use in the pagination.

Here is the code

```
async fetch() {
    let queryShows = query(
        collection(db, "shows"),
        orderBy("ano", "desc"),
        orderBy("nombre", "asc"),
        limit(3)
    );
    const querySnapshotShows = await getDocs(queryShows);
    this.lastDoc = querySnapshotShows.docs[querySnapshotShows.docs.length - 1];
    querySnapshotShows.forEach((doc) => {
        this.shows.push({
            id: doc.id,
            ...doc.data()
        });
    });
},
```

**Comment by wobsoriano** [comment]
It's because you are using the `Firebase Javascript SDK` which only works on client-side that's why it's working on the mounted hook.

What you want to do is to install the [Firebase Admin SDK](https://firebase.google.com/docs/admin/setup), create an API route and from there fetch your data and return the results. Then in your `fetch` method, call the api

**Reply by tomifis** [reply]
I see your point, but when i fetch (using the fetch hook) and without returning the last document cursor it works perfectly, and the content is server side rendered too.

**Reply by wobsoriano** [reply]
Try to put only POJOs in your data. So try to change

```ts
this.lastDoc = querySnapshotShows.docs[querySnapshotShows.docs.length - 1];
```

to

```ts
const lastItem = querySnapshotShows.docs[querySnapshotShows.docs.length - 1]
this.lastDoc = {
   id: lastItem.id,
   ...lastItem.data()
};
```

**Reply by tomifis** [reply]
That doesn´t work. In order to do pagination with firestore you must return a cursor from the last document. The code you share before returns the document object.

**Reply by wobsoriano** [reply]
Ah you're doing pagination. Maybe a simple repro would help

**Reply by tomifis** [reply]
I'm sorry, what do you mean by repro?

**Reply by wobsoriano** [reply]
reproduction. Simple github repro or stackblitz/codesandbox

**Reply by csechuan** [reply]
use firebase admin in API routes. I cant find anywhere to put this piece of code because only api routes / server middlewares are guaranteed in server.
